File: app/ClassScribe.Windows/src/ClassScribe.Core/AsrWorkerProtocol.cs
using System.Text.Json;
using System.Text.Json.Serialization;

namespace ClassScribe.Core;

public static class AsrWorkerProtocol
{
    public const int SupportedVersion = 1;
    public const int MaximumMessageBytes = 1_048_576;
}

[JsonConverter(typeof(AsrMessageTypeJsonConverter))]
public enum AsrMessageType
{
    Hello,
    Ready,
    Heartbeat,
    Progress,
    Start,
    Result,
    RecoverableError,
    TerminalError,
    Cancel,
    Cancelled,
    Shutdown,
}

public sealed class AsrWorkerProtocolException : Exception
{
    public AsrWorkerProtocolException(string message) : base(message) { }
}

public sealed record AsrWorkerEnvelope
{
    [JsonPropertyName("protocolVersion")]
    public int ProtocolVersion { get; init; } = AsrWorkerProtocol.SupportedVersion;
    [JsonPropertyName("attemptID")]
    public SessionAttemptID AttemptID { get; init; } = new();
    [JsonPropertyName("jobID")]
    public Guid JobID { get; init; }
    [JsonPropertyName("messageType")]
    public AsrMessageType MessageType { get; init; }
    [JsonPropertyName("supportedVersions")]
    public IReadOnlyList<int>? SupportedVersions { get; init; }
    [JsonPropertyName("selectedVersion")]
    public int? SelectedVersion { get; init; }
    [JsonPropertyName("sourceReference")]
    public string? SourceReference { get; init; }
    [JsonPropertyName("progress")]
    public double? Progress { get; init; }
    [JsonPropertyName("text")]
    public string? Text { get; init; }
    [JsonPropertyName("code")]
    public string? Code { get; init; }
    [JsonPropertyName("message")]
    public string? Message { get; init; }

    public static AsrWorkerEnvelope Hello(SessionAttemptID attemptID, Guid jobID) => new()
    {
        AttemptID = attemptID,
        JobID = jobID,
        MessageType = AsrMessageType.Hello,
        SupportedVersions = [AsrWorkerProtocol.SupportedVersion],
    };

    public static AsrWorkerEnvelope Ready(
        SessionAttemptID attemptID,
        Guid jobID,
        int protocolVersion = AsrWorkerProtocol.SupportedVersion,
        int selectedVersion = AsrWorkerProtocol.SupportedVersion) => new()
    {
        ProtocolVersion = protocolVersion,
        AttemptID = attemptID,
        JobID = jobID,
        MessageType = AsrMessageType.Ready,
        SelectedVersion = selectedVersion,
    };

    public static AsrWorkerEnvelope Heartbeat(SessionAttemptID attemptID, Guid jobID) => new()
    {
        AttemptID = attemptID,
        JobID = jobID,
        MessageType = AsrMessageType.Heartbeat,
    };

    public static AsrWorkerEnvelope ProgressMessage(
        SessionAttemptID attemptID,
        Guid jobID,
        double progress) => new()
    {
        AttemptID = attemptID,
        JobID = jobID,
        MessageType = AsrMessageType.Progress,
        Progress = progress,
    };

    public static AsrWorkerEnvelope StartMessage(
        SessionAttemptID attemptID,
        Guid jobID,
        string? sourceReference = null) => new()
    {
        AttemptID = attemptID,
        JobID = jobID,
        MessageType = AsrMessageType.Start,
        SourceReference = sourceReference,
    };

    public static AsrWorkerEnvelope Result(
        SessionAttemptID attemptID,
        Guid jobID,
        string? text) => new()
    {
        AttemptID = attemptID,
        JobID = jobID,
        MessageType = AsrMessageType.Result,
        Text = text,
    };

    public static AsrWorkerEnvelope Error(
        SessionAttemptID attemptID,
        Guid jobID,
        bool recoverable,
        string code,
        string message) => new()
    {
        AttemptID = attemptID,
        JobID = jobID,
        MessageType = recoverable ? AsrMessageType.RecoverableError : AsrMessageType.TerminalError,
        Code = code,
        Message = message,
    };

    public byte[] Encode(JsonSerializerOptions? options = null)
    {
        ValidateShape();
        var data = JsonSerializer.SerializeToUtf8Bytes(this, options);
        if (data.Length > AsrWorkerProtocol.MaximumMessageBytes)
        {
            throw new AsrWorkerProtocolException("El mensaje ASR excede el límite de tamaño.");
        }

        return data;
    }

    public static AsrWorkerEnvelope Decode(
        ReadOnlySpan<byte> data,
        JsonSerializerOptions? options = null)
    {
        if (data.Length > AsrWorkerProtocol.MaximumMessageBytes)
        {
            throw new AsrWorkerProtocolException("El mensaje ASR excede el límite de tamaño.");
        }

        var message = JsonSerializer.Deserialize<AsrWorkerEnvelope>(data, options)
            ?? throw new AsrWorkerProtocolException("El mensaje ASR está vacío.");
        message.ValidateShape();
        return message;
    }

    public void ValidateShape()
    {
        if (ProtocolVersion <= 0)
        {
            throw new AsrWorkerProtocolException("La versión del protocolo ASR debe ser positiva.");
        }

        if (AttemptID.SessionID == Guid.Empty
            || AttemptID.Generation <= 0
            || AttemptID.Nonce == Guid.Empty
            || JobID == Guid.Empty)
        {
            throw new AsrWorkerProtocolException("El mensaje ASR no tiene identidad válida.");
        }

        switch (MessageType)
        {
            case AsrMessageType.Hello when SupportedVersions is null
                || SupportedVersions.Count == 0
                || SupportedVersions.Any(version => version <= 0):
                throw new AsrWorkerProtocolException("hello sin versiones soportadas.");
            case AsrMessageType.Ready when SelectedVersion is null || SelectedVersion <= 0:
                throw new AsrWorkerProtocolException("ready sin versión seleccionada.");
            case AsrMessageType.Progress when Progress is null
                || !double.IsFinite(Progress.Value)
                || Progress.Value < 0
                || Progress.Value > 1:
                throw new AsrWorkerProtocolException("progreso fuera de rango.");
            case AsrMessageType.Result when Text is null:
                throw new AsrWorkerProtocolException("result sin texto.");
            case AsrMessageType.RecoverableError or AsrMessageType.TerminalError
                when string.IsNullOrWhiteSpace(Code) || string.IsNullOrWhiteSpace(Message):
                throw new AsrWorkerProtocolException("error sin código o mensaje.");
        }
    }
}

public enum AsrWorkerExitKind
{
    Clean,
    Crashed,
}

public sealed record AsrWorkerExit(AsrWorkerExitKind Kind, int Status);

public enum AsrWorkerTransportFaultKind
{
    LaunchFailed,
    MalformedFrame,
    OversizedFrame,
    UnexpectedEof,
    ReadFailed,
    WriteFailed,
}

public sealed record AsrWorkerTransportFault(
    AsrWorkerTransportFaultKind Kind,
    string? Message = null);

public interface IAsrWorkerTransport
{
    event Action<AsrWorkerEnvelope>? MessageReceived;
    event Action<AsrWorkerExit>? Exited;
    event Action<AsrWorkerTransportFault>? Faulted;
    bool IsTerminated { get; }

    // Implementations must enqueue writes and return without waiting for IPC.
    // Terminate must be able to kill the process independently of that queue.
    void Start(AsrWorkerEnvelope hello);
    void Send(AsrWorkerEnvelope message);
    void Terminate();
}

internal sealed class AsrMessageTypeJsonConverter : JsonConverter<AsrMessageType>
{
    public override AsrMessageType Read(ref Utf8JsonReader reader, Type typeToConvert, JsonSerializerOptions options)
    {
        if (reader.TokenType != JsonTokenType.String)
        {
            throw new JsonException("messageType debe ser un token string.");
        }

        var value = reader.GetString();
        return value switch
        {
            "hello" => AsrMessageType.Hello,
            "ready" => AsrMessageType.Ready,
            "heartbeat" => AsrMessageType.Heartbeat,
            "progress" => AsrMessageType.Progress,
            "start" => AsrMessageType.Start,
            "result" => AsrMessageType.Result,
            "recoverableError" => AsrMessageType.RecoverableError,
            "terminalError" => AsrMessageType.TerminalError,
            "cancel" => AsrMessageType.Cancel,
            "cancelled" => AsrMessageType.Cancelled,
            "shutdown" => AsrMessageType.Shutdown,
            _ => throw new JsonException($"messageType desconocido: {value}"),
        };
    }

    public override void Write(Utf8JsonWriter writer, AsrMessageType value, JsonSerializerOptions options) =>
        writer.WriteStringValue(value switch
        {
            AsrMessageType.Hello => "hello",
            AsrMessageType.Ready => "ready",
            AsrMessageType.Heartbeat => "heartbeat",
            AsrMessageType.Progress => "progress",
            AsrMessageType.Start => "start",
            AsrMessageType.Result => "result",
            AsrMessageType.RecoverableError => "recoverableError",
            AsrMessageType.TerminalError => "terminalError",
            AsrMessageType.Cancel => "cancel",
            AsrMessageType.Cancelled => "cancelled",
            AsrMessageType.Shutdown => "shutdown",
            _ => throw new ArgumentOutOfRangeException(nameof(value)),
        });
}

File: app/MeetingTranscriber/ClassScribeSources/AsrWorkerProtocol.swift
import Foundation

enum AsrWorkerProtocol {
    static let supportedVersion = 1
    static let maximumMessageBytes = 1_048_576
}

enum AsrWorkerProtocolSerializationError: Error, Equatable {
    case messageTooLarge
}

enum AsrWorkerMessageType: String, Codable, Equatable, Sendable {
    case hello
    case ready
    case heartbeat
    case progress
    case start
    case result
    case recoverableError
    case terminalError
    case cancel
    case cancelled
    case shutdown
}

enum AsrWorkerProtocolValidationError: Error, Equatable, LocalizedError, Sendable {
    case invalidProtocolVersion
    case invalidIdentity
    case malformedPayload(String)

    var errorDescription: String? {
        switch self {
        case .invalidProtocolVersion:
            return "La versión del protocolo ASR no es válida."
        case .invalidIdentity:
            return "El mensaje ASR no tiene identidad de intento o job válida."
        case let .malformedPayload(message):
            return "El payload del protocolo ASR es inválido: \(message)"
        }
    }
}

struct AsrWorkerEnvelope: Codable, Equatable, Sendable {
    var protocolVersion: Int
    var attemptID: SessionAttemptID
    var jobID: UUID
    var messageType: AsrWorkerMessageType
    var supportedVersions: [Int]?
    var selectedVersion: Int?
    var sourceReference: String?
    var progress: Double?
    var text: String?
    var code: String?
    var message: String?

    init(
        protocolVersion: Int = AsrWorkerProtocol.supportedVersion,
        attemptID: SessionAttemptID,
        jobID: UUID,
        messageType: AsrWorkerMessageType,
        supportedVersions: [Int]? = nil,
        selectedVersion: Int? = nil,
        sourceReference: String? = nil,
        progress: Double? = nil,
        text: String? = nil,
        code: String? = nil,
        message: String? = nil,
    ) {
        self.protocolVersion = protocolVersion
        self.attemptID = attemptID
        self.jobID = jobID
        self.messageType = messageType
        self.supportedVersions = supportedVersions
        self.selectedVersion = selectedVersion
        self.sourceReference = sourceReference
        self.progress = progress
        self.text = text
        self.code = code
        self.message = message
    }

    static func hello(attemptID: SessionAttemptID, jobID: UUID) -> Self {
        Self(
            attemptID: attemptID,
            jobID: jobID,
            messageType: .hello,
            supportedVersions: [AsrWorkerProtocol.supportedVersion],
        )
    }

    static func ready(
        attemptID: SessionAttemptID,
        jobID: UUID,
        protocolVersion: Int = AsrWorkerProtocol.supportedVersion,
        selectedVersion: Int = AsrWorkerProtocol.supportedVersion,
    ) -> Self {
        Self(
            protocolVersion: protocolVersion,
            attemptID: attemptID,
            jobID: jobID,
            messageType: .ready,
            selectedVersion: selectedVersion,
        )
    }

    static func heartbeat(attemptID: SessionAttemptID, jobID: UUID) -> Self {
        Self(attemptID: attemptID, jobID: jobID, messageType: .heartbeat)
    }

    static func progress(
        attemptID: SessionAttemptID,
        jobID: UUID,
        fraction: Double,
    ) -> Self {
        Self(
            attemptID: attemptID,
            jobID: jobID,
            messageType: .progress,
            progress: fraction,
        )
    }

    static func start(
        attemptID: SessionAttemptID,
        jobID: UUID,
        sourceReference: String? = nil,
    ) -> Self {
        Self(
            attemptID: attemptID,
            jobID: jobID,
            messageType: .start,
            sourceReference: sourceReference,
        )
    }

    static func result(
        attemptID: SessionAttemptID,
        jobID: UUID,
        text: String?,
    ) -> Self {
        Self(
            attemptID: attemptID,
            jobID: jobID,
            messageType: .result,
            text: text,
        )
    }

    static func error(
        attemptID: SessionAttemptID,
        jobID: UUID,
        recoverable: Bool,
        code: String,
        message: String,
    ) -> Self {
        Self(
            attemptID: attemptID,
            jobID: jobID,
            messageType: recoverable ? .recoverableError : .terminalError,
            code: code,
            message: message,
        )
    }

    func validateShape() throws {
        guard protocolVersion > 0 else {
            throw AsrWorkerProtocolValidationError.invalidProtocolVersion
        }
        let zeroUUID = "00000000-0000-0000-0000-000000000000"
        guard attemptID.generation > 0,
              attemptID.sessionID.uuidString.lowercased() != zeroUUID,
              attemptID.nonce.uuidString.lowercased() != zeroUUID,
              jobID.uuidString.lowercased() != zeroUUID
        else {
            throw AsrWorkerProtocolValidationError.invalidIdentity
        }

        switch messageType {
        case .hello:
            guard let supportedVersions,
                  !supportedVersions.isEmpty,
                  supportedVersions.allSatisfy({ $0 > 0 })
            else {
                throw AsrWorkerProtocolValidationError.malformedPayload("hello sin versiones soportadas")
            }
        case .ready:
            guard let selectedVersion, selectedVersion > 0 else {
                throw AsrWorkerProtocolValidationError.malformedPayload("ready sin versión seleccionada")
            }
        case .progress:
            guard let progress, progress.isFinite, (0 ... 1).contains(progress) else {
                throw AsrWorkerProtocolValidationError.malformedPayload("progreso fuera de rango")
            }
        case .result:
            guard text != nil else {
                throw AsrWorkerProtocolValidationError.malformedPayload("result sin texto")
            }
        case .recoverableError, .terminalError:
            guard let code, !code.isEmpty, let message, !message.isEmpty else {
                throw AsrWorkerProtocolValidationError.malformedPayload("error sin código o mensaje")
            }
        case .heartbeat, .start, .cancel, .cancelled, .shutdown:
            break
        }
    }

    func encodedData() throws -> Data {
        try validateShape()
        let data = try JSONEncoder().encode(self)
        guard data.count <= AsrWorkerProtocol.maximumMessageBytes else {
            throw AsrWorkerProtocolSerializationError.messageTooLarge
        }
        return data
    }

    static func decode(_ data: Data) throws -> Self {
        guard data.count <= AsrWorkerProtocol.maximumMessageBytes else {
            throw AsrWorkerProtocolSerializationError.messageTooLarge
        }
        let message = try JSONDecoder().decode(Self.self, from: data)
        try message.validateShape()
        return message
    }
}

enum AsrWorkerExit: Equatable, Sendable {
    case clean(Int32)
    case crashed(Int32)
}

enum AsrWorkerTransportFault: Error, Equatable, Sendable {
    case launchFailed(String)
    case malformedFrame(String)
    case oversizedFrame
    case unexpectedEOF
    case readFailed(String)
    case writeFailed(String)

    var message: String {
        switch self {
        case let .launchFailed(message), let .malformedFrame(message), let .readFailed(message), let .writeFailed(message):
            message
        case .oversizedFrame:
            "El frame ASR excede el límite de tamaño."
        case .unexpectedEOF:
            "El transporte ASR terminó antes de completar un frame."
        }
    }
}

protocol AsrWorkerTransport: AnyObject {
    var onMessage: ((AsrWorkerEnvelope) -> Void)? { get set }
    var onExit: ((AsrWorkerExit) -> Void)? { get set }
    var onFault: ((AsrWorkerTransportFault) -> Void)? { get set }
    var isTerminated: Bool { get }

    /// `start` and `send` must return without waiting for a worker write or
    /// process completion. `terminate` is independent from that write path.
    func start(with hello: AsrWorkerEnvelope)
    func send(_ message: AsrWorkerEnvelope)
    func terminate()
}
